plugins/filter_aws/aws.c

```diff
@@ -370,6 +385,134 @@ static int get_vpc_metadata(struct flb_filter_aws *ctx)
     return ret;
 }
 
+/* Get EC2 instance tag keys from /latest/meta-data/tags/instance.
+ * Initializes ctx->tags_count, ctx->tag_keys and ctx->tag_keys_len.
+ */
+static int get_ec2_tag_keys(struct flb_filter_aws *ctx)
+{
+    int ret;
+    flb_sds_t tags_list = NULL;
+    size_t len = 0;
+    size_t tag_index = 0;
+    size_t tag_start = 0;
+    size_t tag_end = 0;
+    size_t characters_to_copy;
+    size_t i;
+
+    /* get a list of tag keys from the meta data server */
+    ret = get_metadata(ctx, FLB_FILTER_AWS_IMDS_INSTANCE_TAG, &tags_list, &len);
+    if (ret < 0) {
+        if (ret == -2) {
+            flb_warn("[filter_aws] tags not available in the EC2 instance metadata");
+            ctx->tags_count = 0;
+            return 0;
+        }
+        flb_sds_destroy(tags_list);
+        return -1;
+    }
+    /* count number of tag keys and allocate memory */
+    /* it assumes there always is at least one tag */
+    /* starts with 1, because '\n' is a split character */
+    ctx->tags_count = 1;
+    for (i = 0; i < len; i++) {
+        if (tags_list[i] == '\n') {
+            ctx->tags_count++;
+        }
+    }
+    ctx->tag_keys = flb_calloc(ctx->tags_count, sizeof(flb_sds_t*));
+    ctx->tag_keys_len = flb_calloc(ctx->tags_count, sizeof(size_t*));
```

Comment:
after every call that allocates memory you always need an if statement to check if it succeeded and then call `flb_errno()` if it failed. The existing code has many examples.

Reply:
Also, if these are just pointers on the context then you don't need to allocate any memory here

Reply:
So the idea here is to create an array of strings for the tags?

Reply:
Also, you need to protect against running this function twice I think? And then overwriting the previous allocation to these context variables? 

So you add a check like:  `if (ctx->tags_keys != NULL) { flb_free(ctx->tag_keys }`

Reply:
> Also, you need to protect against running this function twice I think? And then overwriting the previous allocation to these context variables?

That's a great point! Thanks!

Based on my understanding, it will repeatedly try to fetch AWS Metadata at each flush until metadata is fully retrieved. However, even if metadata won't be fully retrieved, some information might already be fetched and ready to inject into logs. On one side we need to fetch metadata (free and allocate new ctx->tag_keys and ctx->tag_values), but on the other we need to read from the same list. I just wonder if `cb_aws_filter` might be called concurrently, because if yes, then we need to  have more complexity in order to manage the memory for tags.

> So you add a check like: if (ctx->tags_keys != NULL) { flb_free(ctx->tag_keys }

Okay, so this suggestion will (almost) work if the code isn't executed concurrently. [Almost, because we also need to free items stored inside the array, as tag_keys is just a bunch of pointers.]

Reply:
> Also, you need to protect against running this function twice I think?

Done, https://github.com/fluent/fluent-bit/pull/6065/commits/fb50aa099a508dbf65647d442b47b85178aedc5b

Reply:
Conversation resolved due to lack of new activity. (Feel free to bring it back.)